src/components/data-table/sort-filter-worker.ts

```diff
@@ -1,36 +1,58 @@
 import { expose } from "comlink";
-import { stringCompare, ipCompare } from "../../common/string/compare";
+import Fuse from "fuse.js";
+import memoizeOne from "memoize-one";
+import { ipCompare, stringCompare } from "../../common/string/compare";
 import { stripDiacritics } from "../../common/string/strip-diacritics";
+import { HaFuse } from "../../resources/fuse";
 import type {
   ClonedDataTableColumnData,
   DataTableRowData,
   SortableColumnContainer,
   SortingDirection,
 } from "./ha-data-table";
 
+const fuseIndex = memoizeOne(
+  (data: DataTableRowData[], columns: SortableColumnContainer) => {
+    const searchKeys = new Set<string>();
+    Object.entries(columns)
+      .filter(([, column]) => column.filterable)
+      .forEach(([key, column]) =>
+        searchKeys.add(
+          column.filterKey
+            ? `${column.valueColumn || key}.${column.filterKey}`
+            : key
+        )
+      );
+    return Fuse.createIndex([...searchKeys], data);
+  }
+);
+
 const filterData = (
   data: DataTableRowData[],
   columns: SortableColumnContainer,
   filter: string
 ) => {
   filter = stripDiacritics(filter.toLowerCase());
-  return data.filter((row) =>
-    Object.entries(columns).some((columnEntry) => {
-      const [key, column] = columnEntry;
-      if (column.filterable) {
-        const value = String(
-          column.filterKey
-            ? row[column.valueColumn || key][column.filterKey]
-            : row[column.valueColumn || key]
-        );
-
-        if (stripDiacritics(value).toLowerCase().includes(filter)) {
-          return true;
-        }
-      }
-      return false;
-    })
+
+  if (filter === "") {
+    return data;
+  }
+
+  const index = fuseIndex(data, columns);
+
+  const fuse = new HaFuse(
+    data,
+    { shouldSort: false, minMatchCharLength: 1 },
+    index
   );
+
+  const searchResults = fuse.multiTermsSearch(filter);
+
+  if (searchResults) {
+    return searchResults.map((result) => result.item);
+  }
+
+  return data;
```

Comment:
Does that mean we return all data when there are no results? 🤔

Reply:
If no result, it will be empty array
If no valid search (for example, all term are less than 1 character), it will be undefined

Reply:
```ts
const filteredTerms = minMatchCharLength
  ? terms.filter((term) => term.length >= minMatchCharLength)
  : terms;

if (filteredTerms.length === 0) {
  // If no valid terms are found, return null to indicate no search was performed
  return null;
}
```